### PR TITLE
Feature 106 pscale session based auth

### PR DESCRIPTION
**Pull request by atye**
# Description

Associated driver PR: https://eos2git.cec.lab.emc.com/DevCon/csi-powerscale/pull/312

The above PR adds a MountEndpoint field to the csi-powerscale credential secret so that the driver can retain the original Endpoint after it has been injected with Authorization.

This PR adds support for PowerScale role creation, csi-powerscale sidecar injection, and enables session-based authentication instead of basic authentication when connecting to the backend PowerScale.

Volume creation is successful on 8.x and 9.x clusters and in a multi-cluster configuration. The driver secret can also be updated dynamically with the actual information of a new cluster and re-injected with the sidecar-proxy to obfuscate the new cluster. Volume creation with an Access Zone (non-API endpoint) is also successful.

The latest changes require the latest gopowerscale client so we have a `replace` in the go.mod to account for this. This will need to be updated in the release.

# Issues

List the issues impacted by this PR:

| Issue ID |
| -------- |
|     #106      |

# Checklist:

- [x] I have performed a self-review of my own changes.
```
# make test
docker run --rm -it -v /root/karavi-authorization/policies:/policies/ openpolicyagent/opa test -v /policies/
data.karavi.authz.url.test_get_api_login_allowed: PASS (1.377606ms)
data.karavi.authz.url.test_post_proxy_refresh_token_allowed: PASS (737.596µs)
data.karavi.authz.url.test_get_api_version_allowed: PASS (532.811µs)
data.karavi.authz.url.test_get_system_instances_allowed: PASS (612.343µs)
data.karavi.authz.url.test_get_storagpool_instances_allowed: PASS (540.308µs)
data.karavi.authz.url.test_post_volume_instances_allowed: PASS (601.56µs)
data.karavi.authz.url.test_get_volume_instance_allowed: PASS (586.776µs)
data.karavi.authz.url.test_post_volume_instances_queryIdByKey_allowed: PASS (685.816µs)
data.karavi.authz.url.test_get_system_sdc_allowed: PASS (536.619µs)
data.karavi.authz.url.test_post_volume_add_sdc_allowed: PASS (585.628µs)
data.karavi.authz.url.test_post_volume_remove_sdc_allowed: PASS (542.973µs)
data.karavi.authz.url.test_post_volume_remove_allowed: PASS (901.341µs)
data.karavi.volumes.create.test_small_request_allowed: PASS (1.497396ms)
data.karavi.volumes.create.test_large_request_not_allowed: PASS (316.651µs)
--------------------------------------------------------------------------------
PASS: 14/14
go test -count=1 -cover -race -timeout 30s -short ./...
?       karavi-authorization/cmd/karavictl      [no test files]
ok      karavi-authorization/cmd/karavictl/cmd  20.285s coverage: 65.9% of statements
ok      karavi-authorization/cmd/proxy-server   0.115s  coverage: 15.0% of statements
?       karavi-authorization/cmd/sidecar-proxy  [no test files]
ok      karavi-authorization/cmd/tenant-service 0.063s  coverage: 9.4% of statements
ok      karavi-authorization/deploy     0.102s  coverage: 86.3% of statements
?       karavi-authorization/internal/decision  [no test files]
ok      karavi-authorization/internal/powerflex 9.102s  coverage: 88.9% of statements
ok      karavi-authorization/internal/proxy     6.593s  coverage: 68.3% of statements
ok      karavi-authorization/internal/quota     0.376s  coverage: 90.1% of statements
ok      karavi-authorization/internal/roles     0.053s  coverage: 93.2% of statements
ok      karavi-authorization/internal/tenantsvc 1.625s  coverage: 82.3% of statements
ok      karavi-authorization/internal/token     0.038s  coverage: 90.9% of statements
ok      karavi-authorization/internal/token/jwx 0.058s  coverage: 71.7% of statements
ok      karavi-authorization/internal/web       0.045s  coverage: 23.1% of statements
?       karavi-authorization/pb [no test files]
```

### CODE REVIEW

File: policies/volumes_powerscale_create.rego [hunk 1]
```diff
@@ -39,15 +39,15 @@ deny[msg] {
...
-deny[msg] {
-  count(permitted_roles) == 0
-  msg := sprintf("no roles in [%s] allow the %v Kb request on %s/%s/%s",
-           [input.claims.roles,
-           input.request.volumeSizeInKb,
-           input.systemtype,
-           input.storagesystemid,
-           input.storagepool])
-}
+#deny[msg] {
```
Comment (alikdell) [inline policies/volumes_powerscale_create.rego:42]: Commented code!

File: policies/volumes_powerscale_create.rego [hunk 2]
```diff
@@ -24,7 +24,7 @@ default allow = false
 # a role configured to allow the storage request.
 #
 allow {
-  count(permitted_roles) != 0
+  #count(permitted_roles) != 0
```
Comment (alikdell) [inline policies/volumes_powerscale_create.rego:27]: Commented code!

File: cmd/karavictl/cmd/inject_powerscale.go [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+
+	// Get the config data.
+	configSecData, err := lc.ExtractSecretData(configSecret)
+	if err != nil {
+		lc.Err = fmt.Errorf("getting secret data: %w", err)
+		return
+	}
+
+	// Copy the config data and convert endpoints to localhost:<port>
+	//configSecData = convertEndpoints(configSecData, lc.StartingPortRange)
```
Comment (alikdell) [inline cmd/karavictl/cmd/inject_powerscale.go:87]: Commented code

File: cmd/karavictl/cmd/inject_powerscale.go [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+	secretName := "karavi-authorization-config"
+	authVolume := corev1.Volume{}
+	authVolume.Name = "karavi-authorization-config"
+	authVolume.Secret = &corev1.SecretVolumeSource{
+		SecretName: secretName,
+	}
+	deploy.Spec.Template.Spec.Volumes = append(deploy.Spec.Template.Spec.Volumes, authVolume)
+
+	volumes := deploy.Spec.Template.Spec.Volumes
+	/*for i, v := range volumes {
```
Comment (alikdell) [inline cmd/karavictl/cmd/inject_powerscale.go:153]: Commented code

File: cmd/karavictl/cmd/inject_powerscale.go [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+	secretName := "karavi-authorization-config"
+	authVolume := corev1.Volume{}
+	authVolume.Name = "karavi-authorization-config"
+	authVolume.Secret = &corev1.SecretVolumeSource{
+		SecretName: secretName,
+	}
+	ds.Spec.Template.Spec.Volumes = append(ds.Spec.Template.Spec.Volumes, authVolume)
+
+	volumes := ds.Spec.Template.Spec.Volumes
+	/*for i, v := range volumes {
```
Comment (alikdell) [inline cmd/karavictl/cmd/inject_powerscale.go:257]: Commented code

File: cmd/karavictl/cmd/inject_powerscale.go (hunk 3 above)
Comment (hoppea2) [inline cmd/karavictl/cmd/inject_powerscale.go:87]: should these be removed?

File: cmd/karavictl/cmd/role.go [hunk 6]
```diff
@@ -147,6 +148,11 @@ var GetPowerMaxEndpoint = func(storageSystemDetails System) string {
 	return storageSystemDetails.Endpoint
 }
 
+// GetPowerScaleEndpoint returns the endpoint URL for a PowerMax system
```
Comment (alikdell) [inline cmd/karavictl/cmd/role.go:151]: Comment should be endpoint URL for PowerScale not PowerMax?

File: cmd/proxy-server/main.go [hunk 7]
```diff
@@ -300,6 +300,16 @@ func run(log *logrus.Entry) error {
 			log.WithError(err).Error("main: updating powerscale systems")
 			return
 		}
+		err = powerScaleHandler.UpdateSystems(context.Background(), bytes.NewReader(b), log)
+		if err != nil {
+			log.Errorf("main: failed to update system: %+v", err)
+			return
+		}
+		err = powerScaleHandler.UpdateSystems(context.Background(), bytes.NewReader(b), log)
```
Comment (alikdell) [inline cmd/proxy-server/main.go:308]: Is this duplicate code? Why doing same twice?

File: cmd/karavictl/cmd/role.go [hunk 8]
```diff
@@ -217,6 +223,30 @@ func validatePowerMaxStorageResourcePool(storageSystemDetails System, storageSys
...
+	if err != nil {
+		return fmt.Errorf("powerscale authentication failed: %+v", err)
+	}
+
+	if _, err := c.GetVolumeWithIsiPath(context.Background(), poolQuota.Pool, "", ""); err != nil {
+		return err
+	}
+
+	if int(poolQuota.Quota) < 0 {
+		return errors.New("the specified quota needs to be a positive number")
```
Comment (hoppea2) [inline cmd/karavictl/cmd/role.go:244]: Should the PowerScale accepted quota be 0GB or some other value (-1GB perhaps), just to highlight that quota limits are not enforced by CSM Auth?

**Review by hoppea2 — COMMENTED** [review]
Should the PowerScale accepted quota be 0GB or some other value, as

File: policies/volumes_powerscale_create.rego (hunk 1 above)
Comment (atye) [reply policies/volumes_powerscale_create.rego]: Removed.

File: cmd/karavictl/cmd/inject_powerscale.go (hunk 4 above)
Comment (atye) [reply cmd/karavictl/cmd/inject_powerscale.go]: Removed.

File: cmd/karavictl/cmd/inject_powerscale.go (hunk 5 above)
Comment (atye) [reply cmd/karavictl/cmd/inject_powerscale.go]: Removed.

File: cmd/karavictl/cmd/inject_powerscale.go (hunk 3 above)
Comment (atye) [reply cmd/karavictl/cmd/inject_powerscale.go]: Removed.

File: cmd/karavictl/cmd/role.go (hunk 6 above)
Comment (atye) [reply cmd/karavictl/cmd/role.go]: Fixed.

File: internal/proxy/powerscale_handler.go [hunk 9]
```diff
@@ -199,11 +231,44 @@ func (h *PowerScaleHandler) spoofLoginRequest(w http.ResponseWriter, r *http.Req
 	}
 }
```
[the remark this answers is not in the document]
Comment (atye) [reply internal/proxy/powerscale_handler.go]: Removed this function. It's not used.

File: cmd/karavictl/cmd/role.go (hunk 8 above)
Comment (hoppea2) [reply cmd/karavictl/cmd/role.go]: Thanks. Let's  reach out to Sean so that the documentation is updated for PowerScale

**Review by alikdell — APPROVED** [review]

File: internal/proxy/powerscale_handler.go [hunk 10]
```diff
@@ -216,40 +281,116 @@ func (h *PowerScaleHandler) writeError(w http.ResponseWriter, msg string, code i
...
+	h.log.Debugf("added session cookie to request header: %v", v.sessionCookie)
+	r.Header.Add("X-CSRF-Token", v.csrfToken)
+	h.log.Debugf("added CSRF token to request header: %v", v.csrfToken)
+
+	// Add referrer header
+	r.Header.Add("Referer", v.Endpoint)
+	return nil
+}
+
+func FetchValueIndexForKey(l string, match string, sep string) (int, int, int) {
```
Comment (LJ-Software) [inline internal/proxy/powerscale_handler.go:357]: According to the linting check this should probably start with a lowercase letter:
`internal/proxy/powerscale_handler.go:357:1: exported function FetchValueIndexForKey should have comment or be unexported`
(my bad)

Comment (LJ-Software) [reply internal/proxy/powerscale_handler.go:357]: Youll have to update the one spot in this file where this function is called too

**Review by tdawe — DISMISSED** [review]

File: internal/proxy/powerscale_handler.go (hunk 10 above)
Comment (atye) [reply internal/proxy/powerscale_handler.go]: Fixed.

**Review by LJ-Software — APPROVED** [review]

**Review by alikdell — APPROVED** [review]

**Pull request merged**